[CORE-15021] Cloud Topics: Level Zero GC delete pipeline

oleiman opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

This PR adds a work_queue for delete operation in level_zero_gc, with the general goal of overlapping list & delete. Napkin math suggests that sequential paging list operations will dominate GC latency at low to moderate throughput, so we want to hide any additional delete latency behind the usually more expensive list.

Request concurrency & memory usage of the pipeline are governed by a pair of semaphores - one to limit the memory consumption of outstanding GC-eligible object keys, the other to limit the number of in-flight delete operations going on in the background. 

Since work_queue tasks are processed sequentially, if a particular task blocks waiting for the request semaphore (i.e. before being pushed to the background), subsequent tasks will stay in the queue rather than stacking up on the semaphore. 

Similarly, each iteration of the GC loop will try to page through the entire level_zero/data prefix, queueing up pages of work as it goes. If the page semaphore is exhausted before we reach the last page of list results, the iteration ends and we reenter the loop after a configurable delay, giving the delete worker a chance to burn down some of the outstanding work.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x

## Release Notes

* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

oleiman committed:
ct/l0/gc: Wire up list_delete_worker

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
ct/l0/gc/tests: Misc improvements

- config points
- swap 'deleted' vector<object> for unordered_map<key>
- mutual exclusion for list & delete handlers

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
ct/l0/gc: Each iteration should page through list results

Limited by the the semaphore in list_delete_worker.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman commented on src/v/cloud_topics/level_zero/gc/tests/level_zero_gc_tests.cc:
```diff
@@ -317,3 +354,69 @@ TEST_F(LevelZeroGCMaxEpochTest, EmptySnapshotNonEmptyReport) {
+
+// make sure we make progress when the total size of eligible keys exceeds the
+// in-flight limit
+// i.e. 50 * 500 * len(key)=72 ~= 1.5MiB > 1MiB
```
don't care

oleiman commented on src/v/cloud_topics/level_zero/gc/level_zero_gc.cc:
```diff
@@ -536,26 +631,25 @@ level_zero_gc::try_to_collect() {
+            // NOTE: take the units unsafely. we don't really care about the
+            // memory limit in particular, just don't want to grow unbounded.
+            u.emplace(
+              seastar::consume_units(page_sem_, object_keys_total_bytes));
```
not any different. the keys are already in memory, and we've already gone to the trouble of fetching them. instead of either throwing the work away or pausing until units free up, we can just queue up the work and bounce back out to the control loop.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--15049ac6-a508-4320-8237-77ee73d9caad-->
<!--TEST RESULTS-->
<!--b9b253fa-ac23-4717-af2f-9223406d277f-->
<!--TEST RESULTS-->
<!--fc45d840-b362-4a1f-9201-766df6cf0459-->
<!--TEST RESULTS-->
<!--9ea673f3-7a5f-4834-ad96-b6095ba34160-->
<!--TEST RESULTS-->
<!--4f2f1048-ab31-4aee-9aa7-0265bcdc966c-->
<!--TEST RESULTS-->
<!--eefd5942-7f20-4e65-9e7a-e3e7e0a4bae0-->
<!--TEST RESULTS-->
<!--650b3898-53e4-4675-b188-8bb241b0cb51-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/78192">build#78192</a></summary>


| test_class              | test_method         | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                              |
|:------------------------|:--------------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| MountUnmountIcebergTest | test_simple_remount | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/78192#019b3549-141b-4000-8e63-a1029385314b | FLAKY         | 7/11     | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.1822, p0=0.2693, reject_threshold=0.0100. adj_baseline=0.4531, p1=0.2595, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/78227">build#78227</a></summary>


| test_class                 | test_method                   | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                           |
|:---------------------------|:------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| LogCompactionTxRemovalTest | test_tx_control_batch_removal | null             | integration | https://buildkite.com/redpanda/redpanda/builds/78227#019b3830-76fe-43fa-9b85-c53d76fcd1ee | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0000, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1000, p1=0.3487, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=LogCompactionTxRemovalTest&test_method=test_tx_control_batch_removal |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/78261">build#78261</a></summary>


| test_class                 | test_method          | test_arguments     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                  |
|:---------------------------|:---------------------|:-------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingRandomOpsTest | test_node_operations | {"failures": true} | integration | https://buildkite.com/redpanda/redpanda/builds/78261#019b3eec-c6b9-4a46-9088-a60c811dbe79 | FLAKY         | 19/21    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0027, p0=0.0530, reject_threshold=0.0100. adj_baseline=0.1000, p1=0.3917, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/78427">build#78427</a></summary>


| test_class               | test_method                             | test_arguments                                                            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                                   |
|:-------------------------|:----------------------------------------|:--------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodesDecommissioningTest | test_decommission_status                | null                                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/78427#019b61eb-9936-4e1a-b801-7b00ced276f1 | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0533, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1516, p1=0.1932, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommission_status                |
| NodesDecommissioningTest | test_flipping_decommission_recommission | {"cloud_topic": true, "node_is_alive": false}                             | integration | https://buildkite.com/redpanda/redpanda/builds/78427#019b61ed-27ff-4434-8dac-8472141a0ac3 | FLAKY         | 17/21    | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.0028, p0=0.0000, reject_threshold=0.0100)                                                                    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_flipping_decommission_recommission |
| SIPartitionMovementTest  | test_cross_shard                        | {"cloud_storage_type": 1, "num_to_upgrade": 0, "with_cloud_topics": true} | integration | https://buildkite.com/redpanda/redpanda/builds/78427#019b61ed-2801-4827-bdf0-190d8edcf3ea | FLAKY         | 8/11     | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.0100, p0=0.0043, reject_threshold=0.0100)                                                                    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SIPartitionMovementTest&test_method=test_cross_shard                         |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/78431">build#78431</a></summary>


| test_class                          | test_method                                             | test_arguments                                | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                                                   |
|:------------------------------------|:--------------------------------------------------------|:----------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodesDecommissioningTest            | test_decommissioning_finishes_after_manual_cancellation | {"cloud_topic": true, "delete_topic": false}  | integration | https://buildkite.com/redpanda/redpanda/builds/78431#019b6d0e-4042-4a8f-bddf-2124d9d315f2 | FLAKY         | 9/11     | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.0000, p0=0.0000, reject_threshold=0.0100)                                                                    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_finishes_after_manual_cancellation |
| NodesDecommissioningTest            | test_flipping_decommission_recommission                 | {"cloud_topic": true, "node_is_alive": false} | integration | https://buildkite.com/redpanda/redpanda/builds/78431#019b6d0e-4038-4469-9950-8e81af5668d2 | FLAKY         | 8/11     | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.0041, p0=0.0007, reject_threshold=0.0100)                                                                    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_flipping_decommission_recommission                 |
| WriteCachingFailureInjectionE2ETest | test_crash_all                                          | {"use_transactions": false}                   | integration | https://buildkite.com/redpanda/redpanda/builds/78431#019b6d0e-4038-4469-9950-8e81af5668d2 | FLAKY         | 26/31    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0723, p0=0.1684, reject_threshold=0.0100. adj_baseline=0.2016, p1=0.2482, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all                               |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/78590">build#78590</a></summary>


| test_class                          | test_method    | test_arguments              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                           | test_history                                                                                                                                                     |
|:------------------------------------|:---------------|:----------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudTopicsL0GCTest                 | test_l0_gc     | {"cloud_storage_type": 2}   | integration | https://buildkite.com/redpanda/redpanda/builds/78590#019b946d-3ee3-4c50-ab0e-9d1bc83d5fc0 | FAIL          | 0/11     | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.0000, p0=0.0000, reject_threshold=0.0100) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=CloudTopicsL0GCTest&test_method=test_l0_gc                     |
| CloudTopicsL0GCTest                 | test_l0_gc     | {"cloud_storage_type": 2}   | integration | https://buildkite.com/redpanda/redpanda/builds/78590#019b946e-c1c0-4700-b440-c54059318515 | FAIL          | 0/11     | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.0000, p0=0.0000, reject_threshold=0.0100) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=CloudTopicsL0GCTest&test_method=test_l0_gc                     |
| CloudTopicsL0GCTest                 | test_l0_gc     | {"cloud_storage_type": 1}   | integration | https://buildkite.com/redpanda/redpanda/builds/78590#019b946d-3ede-4eb6-806d-08997d65a593 | FAIL          | 0/11     | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.0000, p0=0.0000, reject_threshold=0.0100) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=CloudTopicsL0GCTest&test_method=test_l0_gc                     |
| CloudTopicsL0GCTest                 | test_l0_gc     | {"cloud_storage_type": 1}   | integration | https://buildkite.com/redpanda/redpanda/builds/78590#019b946e-c1b5-4fa7-97dc-2a6dc95683d5 | FAIL          | 0/11     | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.0000, p0=0.0000, reject_threshold=0.0100) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=CloudTopicsL0GCTest&test_method=test_l0_gc                     |
| WriteCachingFailureInjectionE2ETest | test_crash_all | {"use_transactions": false} | integration | https://buildkite.com/redpanda/redpanda/builds/78590#019b946e-c1b8-4d47-9233-b2656d2e8e76 | FLAKY         | 13/21    | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.1175, p0=0.0060, reject_threshold=0.0100) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/78600">build#78600</a></summary>


| test_class                          | test_method          | test_arguments              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                     |
|:------------------------------------|:---------------------|:----------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| MountUnmountIcebergTest             | test_simple_remount  | {"cloud_storage_type": 1}   | integration | https://buildkite.com/redpanda/redpanda/builds/78600#019b94d9-8fac-4101-a0a6-8dacf5dcaf44 | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.2020, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.4918, p1=0.0011, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount        |
| ShadowLinkingRandomOpsTest          | test_node_operations | {"failures": true}          | integration | https://buildkite.com/redpanda/redpanda/builds/78600#019b94d8-8d0f-45c5-8daf-67c05190a578 | FLAKY         | 18/21    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0503, p0=0.2665, reject_threshold=0.0100. adj_baseline=0.1435, p1=0.4373, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingRandomOpsTest&test_method=test_node_operations    |
| WriteCachingFailureInjectionE2ETest | test_crash_all       | {"use_transactions": false} | integration | https://buildkite.com/redpanda/redpanda/builds/78600#019b94d9-8faa-4512-9787-efd7ab64a5d9 | FLAKY         | 9/11     | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.1160, p0=0.7087, reject_threshold=0.0100. adj_baseline=0.3093, p1=0.1354, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all |
</details>
<!--/body-->

oleiman commented on src/v/cloud_topics/level_zero/gc/level_zero_gc.cc:
```diff
@@ -501,10 +501,23 @@ seastar::future<> level_zero_gc::delete_worker(
-        co_return 0;
+    size_t total_eligible{0};
+    while (page_sem_.available_units() > 0) {
+        auto res = co_await do_try_to_collect();
```
oops, max epoch calculation needs to come out of the loop

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/78427-->
<!--72e4f62b-3d45-46aa-8a9b-97e1ed92c738-->
<!--af64df94-5739-4ad5-a039-99172c9aac87-->
<!--title-->
## Retry command for [Build#78427](https://buildkite.com/redpanda/redpanda/builds/78427)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
skip-redpanda-build
skip-units
skip-rebase
tests/rptest/tests/partition_movement_test.py::SIPartitionMovementTest.test_cross_shard@{"cloud_storage_type":1,"num_to_upgrade":0,"with_cloud_topics":true}
tests/rptest/tests/nodes_decommissioning_test.py::NodesDecommissioningTest.test_flipping_decommission_recommission@{"cloud_topic":true,"node_is_alive":false}
```
<!--/body-->

oleiman commented:
CI Failures:
- https://buildkite.com/redpanda/redpanda/builds/78427 - shutdown timeouts could be related to this PR (we've added a gate & some background work), but both of these tests seem to be flaky on dev.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/78431-->
<!--2656f582-1580-4029-abb9-85f483d8895f-->
<!--bf92f25f-ba53-46c5-95e6-e95228bcd2b4-->
<!--title-->
## Retry command for [Build#78431](https://buildkite.com/redpanda/redpanda/builds/78431)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
skip-redpanda-build
skip-units
skip-rebase
tests/rptest/tests/nodes_decommissioning_test.py::NodesDecommissioningTest.test_decommissioning_finishes_after_manual_cancellation@{"cloud_topic":true,"delete_topic":false}
tests/rptest/tests/nodes_decommissioning_test.py::NodesDecommissioningTest.test_flipping_decommission_recommission@{"cloud_topic":true,"node_is_alive":false}
```
<!--/body-->

dotnwat commented on src/v/cloud_topics/level_zero/gc/level_zero_gc.cc:
```diff
@@ -437,14 +446,27 @@ seastar::future<> level_zero_gc::worker() {
           cd_log.debug,
           "Received error listing objects during L0 GC: {}",
           candidate_objects.error());
         co_return std::unexpected(collection_error::service_error);
```
based on the comment below

> / TODO: fairly naive approach to caching the token. if the request
>    // fails, we start over,

what does start over mean? does it mean that here in the error case ("!candidate_objects.has_value()") that we should reset the token?

oleiman replied:
yeah, it looks like I misplaced the optional::reset while splitting out this commit from a later one.

dotnwat commented on src/v/cloud_topics/level_zero/gc/level_zero_gc.cc:
```diff
@@ -575,26 +615,25 @@ level_zero_gc::try_to_collect() {
+            // NOTE: take the units unsafely. we don't really care about the
+            // memory limit in particular, just don't want to grow unbounded.
+            u.emplace(
+              seastar::consume_units(page_sem_, object_keys_total_bytes));
```
iiuc we never block on page_sem? if we never block on it, what's it's purpose?

oleiman replied:
just a convenient interface for accounting I guess. note that at the top of try_to_collect we'll return 0 if page_sem is exhausted, so effectively we "block" on the semaphore by backing off in the outer loop

dotnwat commented on src/v/cloud_topics/level_zero/gc/level_zero_gc.cc:
```diff
@@ -444,8 +457,32 @@ seastar::future<> level_zero_gc::worker() {
+    const auto num_eligible = eligible_objects.size();
+
+    auto res = co_await storage_->delete_objects(
+      &delete_as_, std::move(eligible_objects));
```
why do we need a new abort source separate from the one we already have?

oleiman replied:
iiuc the existing abort source gets blown away on every GC iteration, whereas this one is held onto by background work that's not strictly synchronous w/ the driver loop. Didn't want to mess with it, but I think that speaks to the broader point about factoring some of this out.

dotnwat commented on src/v/cloud_topics/level_zero/gc/level_zero_gc.cc:
```diff
@@ -575,26 +615,25 @@ level_zero_gc::try_to_collect() {
+            u.emplace(
+              seastar::consume_units(page_sem_, object_keys_total_bytes));
+        }
+        delete_worker_.submit([this,
```
i wonder if ssx/actor.h would work here? in start create a new actor. in pause/stop you'd destroy the current actor. but similar simplification for work queue might exist?

oleiman replied:
> ssx/actor

first I'm hearing of it, but at a quick glance it looks pretty close.

dotnwat commented on src/v/cloud_topics/level_zero/gc/level_zero_gc.h:
```diff
@@ -267,10 +273,25 @@ class level_zero_gc {
     seastar::future<std::expected<size_t, collection_error>> try_to_collect();
+    seastar::future<std::expected<size_t, collection_error>>
+    do_try_to_collect(std::optional<cluster_epoch>&);
+    std::optional<ss::sstring> continuation_token_;
```
instead of having new work queue, semaphores, gates, abort sources all within this existing class, should we instead factor out paging/deletion into separate "workers" that are self-contained with a simple interface so that level_zero_gc is managing them at a higher level of abstraction (e.g. start/pause/stop)?

oleiman replied:
> instead factor out paging/deletion

yeah maybe so. 

> separate "workers" that are self-contained with a simple interface

Not sure I follow "separate workers". I rather like the work queue itself as a primitive source of backpressure, so in my mind we could just factor this new code into a separate class basically as is and expose some single-method async API to the GC loop. 

Did you have something more granular in mind?

oleiman replied:
> factor this new code

or maybe this is just ssx::actor

dotnwat replied:
> Not sure I follow "separate workers". I rather like the work queue itself as a primitive source of backpressure, so in my mind we could just factor this new code into a separate class basically as is and expose some single-method async API to the GC loop.

Right, nothing wrong with keeping the work queue, I'm talking about hiding the other complexity (semaphores, abort sources, gates, etc...)

oleiman replied:
cool cool. yeah will do.

oleiman replied:
done. if ssx::actor is the better choice it'll be a bit easier to swap that in.

oleiman commented:
[force push](https://github.com/redpanda-data/redpanda/compare/c1df2d68d5d20113109571f56d8f02f26af09a5e..21b123c023d48f8da3bf0975c719c630bf43f73c) rebase dev

oleiman committed:
ct/l0/gc: list_delete_worker

In general, we expect sequential list operations to dominate GC latency.
With that in mind, this commit introduces list_delete_worker to manage and
dispatch outstanding pages of GC-eligible object keys, the idea being that the
_processing_ of a page of objects should not block grabbing the next one.

This worker includes a "page semaphore" to limit the total volume of object
keys held in memory at once. Precise accounting is not the goal here.
We just want to avoid growing the backlog of delete tasks unbounded.

Delete requests are dispatched sequentially but execute in the background.
Another semaphore limits the number of background delete requests in flight
at one time.

Note that splitting a given remote::delete_objects payload into API-sized
batches is handled internally to cloud_io::remote, so we may want to rejigger
the concurrency in this layer a bit.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

oleiman committed:
ct/l0/gc: Add optional continuation token to list_objects

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/78590-->
<!--0b8cc629-cb75-4be9-b26c-48a514e56981-->
<!--1a0a3a28-b6b1-4f36-aeda-e2a66c8a5178-->
<!--145ac599-e8f6-4020-b416-f1cf1c93500e-->
<!--11db68b7-a5eb-4718-b08c-bec5f20e2ad4-->
<!--178f857b-057f-4999-86b4-3d15d75f5ea1-->
<!--title-->
## Retry command for [Build#78590](https://buildkite.com/redpanda/redpanda/builds/78590)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
skip-redpanda-build
skip-units
skip-rebase
tests/rptest/tests/cloud_topics/l0_gc_test.py::CloudTopicsL0GCTest.test_l0_gc@{"cloud_storage_type":1}
tests/rptest/tests/cloud_topics/l0_gc_test.py::CloudTopicsL0GCTest.test_l0_gc@{"cloud_storage_type":2}
tests/rptest/tests/write_caching_fi_e2e_test.py::WriteCachingFailureInjectionE2ETest.test_crash_all@{"use_transactions":false}
```
<!--/body-->

dotnwat commented on src/v/cloud_topics/level_zero/gc/level_zero_gc.cc:
```diff
@@ -15,13 +15,168 @@
+                // unbounded.
+                u.emplace(seastar::consume_units(page_sem_, keys_total_bytes));
+            }
+            worker_.submit([this,
+                            o = std::move(objects),
+                            u = std::move(u).value()]() mutable {
+                return do_delete_objects(std::move(o), std::move(u));
```
it's unclear to me what value the worker provides if the worker turns around and backgrounds the work you gave it. can you clarify its role?

if it because you want to be able to block on the delete_sem, then you could acquire the delete_sem in the background. that is, have many inflight background tasks that each acquire the delete_sem.

that would mean that you'd want to probably limit the number of background tasks that are created (i.e. limit the producer/paging side), but presumably that already exists otherwise the worker's queue would fill up (maybe that is the `!delete_worker_->has_capacity()`)?

oleiman replied:
> want to be able to block on the delete_sem

yes, that

> what value the worker provides...[vs having] many inflight background tasks that each acquire the delete_sem

just erring on the side of fewer background fibers I guess. it feels subjectively better to stack pending work up on a queue we control vs having an arbitrary number of blocked tasks in the background, but I'm not sure there's an empirical reason to prefer one or the other.

I think we had an issue in tiered storage where we pushed some work into the background, wrongly assuming that some higher level concurrency control would supply enough back pressure. But maybe blocking on delete_sem is good enough for this use case.

dotnwat commented on src/v/cloud_topics/level_zero/gc/level_zero_gc.cc:
```diff
@@ -607,9 +607,27 @@ seastar::future<> level_zero_gc::worker() {
+    std::optional<cluster_epoch> max_gc_epoch;
+    size_t total_eligible{0};
+    while (delete_worker_->has_capacity()) {
+        auto res = co_await do_try_to_collect(std::ref(max_gc_epoch));
```
instead of introducing an input/output reference parameter, can we pull the max_gc_epoch query out into this function and pass it in as a value? from the looks of it do_try_collect won't recompute it, so a single query here reused across invocations would be equivalent?

oleiman replied:
> single query here reused across invocations would be equivalent?

fair enough. i assumed that putting the epoch after the LIST request was deliberate to avoid wasting an expensive computation, but I suppose reusing it across several pages is a good enough hedge against LIST failure.

dotnwat replied:
yeh makes sense. i'm fine with whatever, just always a fan of avoiding input/output parameters when possible.
